jsonschema-core.md

```diff
@@ -895,49 +895,44 @@ by other parties.
 ### Base IRI, Anchors, and Dereferencing
 
 To differentiate between schemas in a vast ecosystem, schemas are identified by
-[IRI](#rfc3987), and can embed references to other schemas by specifying their
-IRI.
+[absolute IRIs](#rfc3987) (without fragments) and can embed references to other
+schemas by specifying their respective IRIs. When comparing IRIs for the
+purposes of resource identification, implementations SHOULD first follow the IRI
+normalization procedures defined in [RFC 3987](#rfc3987), section 5.3.
 
-Several keywords can accept a relative [IRI-reference](#rfc3987), or a value
-used to construct a relative IRI-reference. For these keywords, it is necessary
+Several keywords can accept a relative [IRI reference](#rfc3987), or a value
+used to construct a relative IRI reference. For these keywords, it is necessary
 to establish a base IRI in order to resolve the reference.
 
 #### The `$id` Keyword {#id-keyword}
 
-The `$id` keyword identifies a schema resource with its [canonical](#rfc6596)
-IRI.
+The `$id` keyword identifies a schema resource. The value for this keyword MUST
+be a string, and MUST represent a valid [IRI reference](#rfc3987) (without a
+fragment).
+
+When the value of this keyword is resolved against the current base IRI, the
+resulting absolute IRI then serves as the identifier for the schema resource and
+as a base IRI for relative IRI references in keywords within that schema
+resource and for nested schema resources, in accordance with [RFC 3987 section
+6.5](#rfc3987) and [RFC 3986 section 5.1.1](#rfc3986) regarding base IRIs
+embedded in content.
 
 Note that this IRI is an identifier and not necessarily a network locator. In
 the case of a network-addressable URL, a schema need not be downloadable from
 its canonical IRI.
 
-If present, the value for this keyword MUST be a string, and MUST represent a
-valid [IRI-reference](#rfc3987). This IRI-reference SHOULD be normalized, and
-MUST resolve to an [absolute-IRI](#rfc3987) (without a fragment).
-
-The resulting absolute-IRI serves as the base IRI for relative IRI-references in
-keywords within the schema resource, in accordance with [RFC 3987 section
-6.5](#rfc3987) and [RFC 3986 section 5.1.1](#rfc3986) regarding base IRIs
-embedded in content.
-
-The presence of `$id` in a subschema indicates that the subschema constitutes a
-distinct schema resource within a single schema document. Furthermore, in
-accordance with [RFC 3987 section 6.5](#rfc3987) and [RFC 3986 section
-5.1.2](#rfc3986) regarding encapsulating entities, if an `$id` in a subschema is
-a relative IRI-reference, the base IRI for resolving that reference is the IRI
-of the parent schema resource. Note that an `$id` consisting of an empty IRI or
```

Comment:
This needs to stay. As much as I'd like to remove the concept of embedded schemas from the spec, I don't think that was your intention.

Reply:
> The first sentence in this paragraph is just restating the first sentence of the section, "The $id keyword identifies a schema resource."

I see what you mean, but I don't think readers will make the connection that it's not just an identifier for the root schema, but also away to embed a schema resource somewhere inside the root schema. Those who do make the connection will be looking for something else to confirm that connection is correct and won't find it.

> The second sentence here is replaced by the new text on lines 913-918.

Those two things don't say the same thing. The first refers to base IRIs embedded in content and the second refers to encapsulating entities. Those are treated as two different cases in RFC 3986/7, so I think we should treat them as two different cases as well.

I think the spec is better with that paragraph included.

Reply:
Previous text:

> The presence of `$id` in a subschema indicates that the subschema constitutes a
distinct schema resource within a single schema document. ... if an `$id` in a subschema is a relative IRI-reference, the base IRI for resolving that reference is the IRI of the parent schema resource.

New text:

> The `$id` keyword identifies a schema resource. ... When the value of [`$id`] is resolved against the current base IRI, the resulting absolute IRI then serves as the identifier for the schema resource and as a base IRI for relative IRI references in keywords within that schema resource and for nested schema resources...

These texts are saying the same things, specifically:

- `$id` identifies and declares a schema resource
- the value of `$id` is resolved against whatever comes from its parent

Maybe we need to define "current base IRI" better, but I'm almost sure it's defined elsewhere.

As I mentioned [here](https://github.com/json-schema-org/json-schema-spec/pull/1537#issuecomment-2408016806), the resolution of `$id` is somewhat recursive, which makes it hard to explain.

I'll have another go at this, but what's here needs improvement.

Reply:
> for nested schema **_resources_**

I missed that last word. I thought this was talking about subschemas, not embedded schema resources. I think using the term "embedded" rather than "nested" is important to be consistent with how we talk about this concept throughout the spec and in discussion. Using a different word primed my brain to think you were explicitly not talking about embedded schema resources.

So, I think the new text works if the RFC reference mentions both "base IRIs embedded in content" _and_ "encapsulating entities".

To summarize, my suggestions are now
* Make it more explicit that `$id` in a subschema creates an embedded schema resource
* Use "embedded" instead of "nested"
* Reference "encapsulated entities" as well as "base IRIs embedded in content".

If we can do those things, I'm comfortable with the removed paragraph staying removed.

Reply:
> Reference "encapsulated entities" as well as "base IRIs embedded in content".

I'm not sure where you're meaning for this.

Reply:
Looks like you figured out what I meant.

> in accordance with [RFC 3987 section
6.5](#rfc3987) and [RFC 3986 section 5.1.1](#rfc3986) regarding base IRIs
embedded in content.

:point_up: That was what I wanted.

Reply:
For example,

> in accordance with [RFC 3987 section 6.5](#rfc3987) and [RFC 3986 section 5.1.1](#rfc3986) regarding base IRIs embedded in content ***and [RFC 3986 section 5.1.2](#rfc3986) regarding encapsulating entities.***

Section 5.1.1 (base IRIs embedded in content) covers base IRIs.
Section 5.1.2 (encapsulating entities) covers embedded schema resources.

Both need to be referenced.